app/src/ble.c

```diff
@@ -56,19 +56,18 @@ enum advertising_type {
 #define CURR_ADV(adv) (adv << 4)
 
 #define ZMK_ADV_CONN_NAME                                                                          \
-    BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_ONE_TIME, BT_GAP_ADV_FAST_INT_MIN_2, \
-                    BT_GAP_ADV_FAST_INT_MAX_2, NULL)
+    BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_ONE_TIME | BT_LE_ADV_OPT_USE_NAME,   \
+                    BT_GAP_ADV_FAST_INT_MIN_2, BT_GAP_ADV_FAST_INT_MAX_2, NULL)
 
 static struct zmk_ble_profile profiles[ZMK_BLE_PROFILE_COUNT];
 static uint8_t active_profile;
 
 #define DEVICE_NAME CONFIG_BT_DEVICE_NAME
 #define DEVICE_NAME_LEN (sizeof(DEVICE_NAME) - 1)
 
-BUILD_ASSERT(DEVICE_NAME_LEN <= 16, "ERROR: BLE device name is too long. Max length: 16");
+BUILD_ASSERT(DEVICE_NAME_LEN <= 27, "ERROR: BLE device name is too long. Max length: 27");
```

Comment:
Sorry. As I think on this... I believe we limited this further because longer names would cause our advertising data to be too large, when combined with our services listed. So even if the name pieces allow this. I'm not sure we can actually use longer names.

Did you test with a longer name?

Reply:
Good spot, I tested it with an 18 char name, it worked and i called it good, I've just gone back to it, I can get a 20 char name to work but no longer. The characteristic doesn't seem to accept the write if it's longer

Reply:
Happy to cap it at the original 16 if you want. It's possible my laptop is accepting the overlength advertising data when it shouldnt be

Reply:
> Happy to cap it at the original 16 if you want. It's possible my laptop is accepting the overlength advertising data when it shouldnt be

See https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/adv.c#L187

By default, asking Zephyr to advertise the name will put the name in the SD data instead of the AD packet itself, which is likely why this is working for longer names.

Two years ago or so, having name in the SD meant some hosts failed to display a name when pairing. We can test this again *now* and see if that's still a problem or not. SD certainly would be better, if it works.

Reply:
IIRC, it was macOS at the time that wouldn't show a name if it was in SD

Reply:
Alternatively it looks like we can force it to put the name in AD and restrict to the original 16 chars

Reply:
> Alternatively it looks like we can force it to put the name in AD and restrict to the original 16 chars

I'd vote for this for now. Can we do that for names that come in from a write to the GAP characteristic though?

Reply:
Just tested it and the new name is properly reflected in AD after writing to the GAP characteristic

Reply:
of note is after pairing android doesn't reflect the new name, it's got some local name caching because there's a mechanism where you can give it a name that stays on the host device to refer to it